### PR TITLE
S4581378-Siamese network for ISIC 2020 Kaggle Dataset

### PR DESCRIPTION
**Pull request by Mrath10**
The following pull request contains an implementation for a siamese network to classify melanoma for the 256x256 ISIC 2020 melanoma project. the project contains a readme for usage, relevant links and all findings.

### CODE REVIEW

**Comment by hanemma7moud** [comment]
### This is an initial inspection, no action is required at this point
Recognition Problem : total : **20**

Solves problem: The solution is well-suited to the problem (**5**).
+ The model performs reasonably very well overall: Testing accuracy: 83.76%, testing sensitivity: 75.86% , and testing Specificity: 83.93%
+ The t-SNE and PCA plots suggest partial feature separation
+ Contrastive loss is utilized, which is suitable for the problem.
+ Class imbalance is addressed through data augmentation and resampling, ensuring better representation of minority cases

Implementation functions as intended (**3**)

Good design: Well-designed (**1**)

Commenting: Clear and sufficient comments throughout the code. (**1**)

Difficulty: Hard (**10**)

Note:
-The t-SNE and PCA plots (both at epoch 5 and the test set) show some clustering but not distinct separation between benign and malignant classes. In particular, there seems to be significant overlap between the two classes, suggesting that the embeddings are not yet fully separable.
- The AUROC and accuracy curves for training and validation show some improvement, but the validation AUROC fluctuates and even declines over epochs, which might indicate overfitting or insufficient feature discrimination.
- The loss curves suggest some progress in learning, but the presence of overfitting past epoch 5 and the fluctuating validation metrics imply that the model may not be generalizing well, a likely result of insufficient feature separation.
- The classification report also indicates that the model struggles with low precision for the minority class (malignant), which suggests that it produces a high number of false positives.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. Has data splits.</td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard.  Siamese network (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. 1</td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-1</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Good work!
